File: iwrap/generators/wrapper_generators/cpp_wrapper/resources/src/iwrap_tools.cpp
#include <fstream>
#include <sstream>
#include <cstring>
#include <cctype>

#include "iwrap_tools.h"
#include "serialization_tools.h"

int read_input(const char* file_name, ids_description_t db_entry_desc_array[], int array_expected_size)
{
    ifstream fin;
    int array_read_size = -1;

    if (array_expected_size < 1)
        return 0;

    fin.open(file_name);
    fin.ignore(INT_MAX, '\n'); // skip line " === Arguments ===="
    fin.ignore(INT_MAX, '\n'); // skip line " Length:"
    read_data(&fin, &array_read_size);

    if ( array_expected_size != array_read_size)
    {
        printf("ERROR: Expected and read number of arguments differs ( %d  vs  %d )", array_expected_size, array_read_size);
        return -1;
    }

    for (int i=0; i < array_expected_size; i++)
    {
        read_data(&fin, &db_entry_desc_array[i]);
    }
    
    fin.close();

    return 0;
}

void read_code_parameters( char** xml_string)
{
    FILE *f = fopen("code_parameters.xml", "rb");
    fseek(f, 0, SEEK_END);
    long fsize = ftell(f);
    fseek(f, 0, SEEK_SET);

    *xml_string = (char *)malloc(fsize + 1);
    fread(*xml_string, fsize, 1, f);
    fclose(f);

    (*xml_string)[fsize] = 0;
}

int write_output(const char* file_name, int status_code, char* status_message)
{
    ofstream fout;
    int str_len;

    fout.open(file_name);
    write_data(&fout, status_code);

    if ( status_message != NULL)
    {
        str_len = strlen(status_message) ;
        write_data(&fout, str_len);
        write_data(&fout, status_message);
    }
    else
    {
        write_data(&fout, 0);
        write_data(&fout, "");
    }

    fout.close();
    return 0;
}

void handle_status_info(int status_code, char* status_message, const char* actor_name, const char* method_name)
{
    const char* NO_MSG = "<No diagnostic message>";
    if(status_message == NULL)
    {
        int msg_size = strlen(NO_MSG) + 1;
        status_message = (char*) malloc(msg_size);
        strcpy(status_message, NO_MSG);
    }

    if(status_code !=0) {
      printf("---Diagnostic information returned from *** %s/%s ***:---\n", actor_name, method_name);
      printf("-------Status code    : %d\n", status_code);
      printf("-------Status message : %s\n", status_message);
      printf("---------------------------------------------------------\n");
      }
}

void release_status_info(char* status_message)
{
    if(status_message != NULL)
        free(status_message);
}

void convert_status_info(std::string in_status_msg, char** out_status_msg)
{

    //status message conversion
    int size = in_status_msg.length();

    *out_status_msg = (char*) malloc(size + 1);
    strcpy(*out_status_msg, in_status_msg.c_str());
}

IdsNs::IDS* init_db(ids_description_t* db_entry_desc)
{
    IdsNs::IDS* db_entry = new IdsNs::IDS(db_entry_desc->uri, "r");
    return db_entry;
}

int open_db(IdsNs::IDS* db_entry, ids_description_t* db_entry_desc)
{
    return 0;
}



void close_db(IdsNs::IDS* db_entry)
{
    db_entry->close();
}

File: iwrap/generators/wrapper_generators/java_wrapper/resources/src/iWrapTools.java
import imasjava.imas;
import imasjava.wrapper.LowLevel;

import java.io.*;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.Arrays;

public class iWrapTools{

    static String readFile(String path)
            throws IOException
    {
        byte[] encoded = Files.readAllBytes(Paths.get(path));
        return new String(encoded, StandardCharsets.UTF_8);
    }

    static public IDSDescription[] read_input(String file_name, int array_expected_size) throws IOException {

        int readSize = -1;
        IDSDescription idsDescriptions[];

        if (array_expected_size < 1)
        {
            idsDescriptions = new IDSDescription[0];
            return idsDescriptions;
        }

        FileReader fr = new FileReader(file_name);
        BufferedReader br = new BufferedReader(fr);
        String line;

        line = br.readLine();  // skip line " === Arguments ===="
        line = br.readLine();  // skip line " Length:"
        line = br.readLine();  //length
        readSize = Integer.parseInt(line);

        if ( array_expected_size != readSize)
        {
            throw new IOException("ERROR: Expected and read number of arguments differs ( " +  array_expected_size + " vs " + readSize);
        }

        idsDescriptions = new IDSDescription[readSize];
        for (int i=0; i < idsDescriptions.length; i++)
        {
            IDSDescription idsDescription = new IDSDescription();

            idsDescription.read(br);
            idsDescriptions[i] = idsDescription;
        }

        br.close();
        fr.close();

        return idsDescriptions;
    }

    public static String read_code_parameters() throws IOException {
        String xmlString=iWrapTools.readFile("code_parameters.xml");
        return xmlString;
    }

    static public void write_output(String fileName, int statusCode, String statusMessage) throws IOException
    {
        BufferedWriter writer = new BufferedWriter(new FileWriter(fileName, true));
        writer.write("" + statusCode);
        writer.newLine();

        if ( statusMessage != null )
        {
            int str_len =  statusMessage.length();
            writer.write("" + str_len);
            writer.newLine();
            writer.write(statusMessage);
            writer.newLine();
        }
        else
        {
            writer.write("0");
            writer.newLine();
            writer.write("");
            writer.newLine();
        }

        writer.close();
    }

    static public void handle_status_info(int status_code, String status_message, String actor_name, String method_name)
    {
        String NO_MSG = "<No diagnostic message>";
        if(status_message == null)
            status_message = NO_MSG;

        if(status_code !=0) {
          System.out.println("---Diagnostic information returned from *** " + actor_name + " / " + method_name + "***:---\n");
          System.out.println("-------Status code    : " + status_code + "\n");
          System.out.println("-------Status message : " + status_message + "\n" );
          System.out.println("---------------------------------------------------------\n");
        }

    }

    static public void open_db(IDSDescription idsDescription) throws Exception {

        int idx = -1;

        if (idsDescription.backend_id == LowLevel.MEMORY_BACKEND)
            return;

        idx = imas.openEnv(idsDescription.pulse,
                idsDescription.run,
                idsDescription.user,
                idsDescription.database,
                idsDescription.version,
                idsDescription.backend_id);

        idsDescription.idx = idx;
    }

    static public void close_db(IDSDescription idsDescription) {

            if (idsDescription.backend_id == LowLevel.MEMORY_BACKEND)
                return;

            try{
                imas.close(idsDescription.idx);
            } catch (Exception ex){
                // Nothing to do here... It is just cleaning operation.
            }
    }

}
